packages/components/src/utils/useMeasuredElement.ts

```
@@ -29,22 +29,15 @@ import { useResize } from './useResize'
 
 function measureElement(element?: HTMLElement | null) {
   if (!element) {
-    return {
-      bottom: 0,
-      height: 0,
-      left: 0,
-      right: 0,
-      top: 0,
-      width: 0,
-    }
+    return new DOMRect()
```

DOMRect needs to be imported to work in our Jest & Gatsby CI runs.

---

imported from where, though? It's [browser global](https://developer.mozilla.org/en-US/docs/Web/API/DOMRect/DOMRect).

---

Less elegant but this would work and keep the object strongly typed. 

```jsx

function measureElement(element?: HTMLElement | null): DOMRect {
  if (!element) {
    return {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      toJSON: () => undefined,
      top: 0,
      width: 0,
      x: 0,
      y: 0,
    }
  }

  return element.getBoundingClientRect()
}```

For better or worse Gatsby's SSR build has pushed us into a few cases like this where we can't _assume_ the browser APIs are available.

Definitely open to other ideas if you have any :)

---

I don't like that this would fail `instanceof` checks and might not yield the same key order...

---

it seems like attempting to measure an element in an SSR context should return an error token rather than an all-zeros DOMRect

---

An error token instead of the all-zeros DOMRect would work but would require some refactoring at the call sites. It's not so much for SSR context... this hook was actually designed for [use with a `useState`-style ref](https://github.com/looker-open-source/components/blob/main/packages/components/src/Form/Inputs/RangeSlider/RangeSlider.tsx#L211), so `element` could be `null` if the element it represents is mounted or unmounted after or before the component calling the hook. Ironically, I don't think it's being used that way anywhere, but I did add [one `usage` where I pass `null` when I wanted to turn measuring off](https://github.com/looker-open-source/components/blob/main/packages/components/src/utils/useWindow.tsx#L177). Probably a separate `disabled` prop would be better there.